contracts/XorShift128Plus.cdc

```diff
@@ -0,0 +1,95 @@
+import Crypto
+
+/// Defines a xorsift128+ pseudo random generator as a resource
+///
+access(all) contract XorShift128Plus {
+
+    access(all) let StoragePath: StoragePath
+    access(all) let PublicPath: PublicPath
+
+    /// While not limited to 128 bits of state, this PRG is largely informed by XORShift128+
+    ///
+    access(all) resource PRG {
+
+        // The states below are of type Word64 (instead of UInt64) to prevent overflow/underflow as state evolves
+        //
+        access(all) var state0: Word64
+        access(all) var state1: Word64
+
+        init(sourceOfRandomness: [UInt8]) {
+            pre {
+                sourceOfRandomness.length == 32: "Expecting 32 bytes as sourceOfRandomness"
+            }
+            // Convert the seed bytes to two Word64 values for state initialization
+            let segment0: Word64 = XorShift128Plus.bigEndianBytesToWord64(bytes: sourceOfRandomness, start: 0)
+            let segment1: Word64 = XorShift128Plus.bigEndianBytesToWord64(bytes: sourceOfRandomness, start: 8)
+            let segment2: Word64 = XorShift128Plus.bigEndianBytesToWord64(bytes: sourceOfRandomness, start: 16)
+            let segment3: Word64 = XorShift128Plus.bigEndianBytesToWord64(bytes: sourceOfRandomness, start: 24)
+
+            self.state0 = segment0 ^ segment1
+            self.state1 = segment2 ^ segment3
+        }
+
+        /// Advances the PRG state and generates the next UInt64 value
+        /// See https://arxiv.org/pdf/1404.0390.pdf for implementation details and reasoning for triplet selection.
+        /// Note that state only advances when this function is called from a transaction. Calls from within a script
+        /// will not advance state and will return the same value.
+        ///
+        /// @return The next UInt64 value
+        ///
+        access(all) fun nextUInt64(): UInt64 {
+            var a: Word64 = self.state0
+            let b: Word64 = self.state1
+
+            self.state0 = b
+            a = a ^ (a << 23) // a
+            a = a ^ (a >> 17) // b
+            a = a ^ b ^ (b >> 26) // c
+            self.state1 = a
+
+            let randUInt64: UInt64 = UInt64(a + b)
```

Comment:
I think the original code is the correct version, because `UInt64` additions on Cadence must not overflow. I asked in my previous comment to make sure the addition is made as `Word64`

```suggestion
            let randUInt64: UInt64 = UInt64(Word64(a) + Word64(b))
```